### DISCUSSION TITLE
Stale corfu pop-up after comint-previous-matching-input-from-input

### DISCUSSION THREAD

**Discussion by egr95** [minad/corfu#468] [Q&A] [answered]
The emacs inferior shell offers great tricks for access one's command history, so for example, I use the following mapping, which lets me type out the first few characters of a command that I have used in the past, and quickly autocomplete it by pressing `M-p`:
```
(define-key comint-mode-map (kbd "M-p") 'comint-previous-matching-input-from-input)
```

So for example, suppose I open an inferior shell and run `python script.py --args`. Then a few commands later I can just type "py" and press `M-p`, and the rest of my command will populate the prompt. With corfu autocompletion active, after typing "py", the usual corfu pop-up shows up correctly, suggesting completions like "python", but my issue is that once I hit `M-p`, the rest of my command populates but I am left with a stale corfu pop-up that simply says "No match" and doesn't go away after subsequent keypresses. This is annoying because I often want to pull up an earlier command and then add some modifications, and it would be nice to have all the usual benefits of corfu when doing this.

I have solved this issue by adding the following hook in my config, but it seems like this might be something that could be better addressed directly in the package?
```
(defun my-corfu-quit-wrapper (&rest _args)
  "Wrapper function to call `corfu-quit` and ignore any arguments."
  (corfu-quit))

(advice-add 'comint-previous-matching-input-from-input :after 'my-corfu-quit-wrapper)
```

Thanks!

**Comment by minad** [comment, marked as answer]
Hi @egr95! You are probably aware, that the "No match" popup can be prevented by setting `corfu-quit-no-match=t`. Furthermore there is the setting `corfu-quit-no-match=separator` (the default), but I doubt that this works in this case, since `comint-previous-matching-input-from-input` already inserted spaces? Note that all these settings can be set buffer-locally only in Comint buffers. Still, I believe your advice might be indeed the best solution here. Another option to consider would be to disable auto completion in comint mode. Does this help?

**Reply by minad** [reply]
Two other solutions which I use instead of `comint-previous-matching-input-from-input` (or the Eshell equivalent) are `consult-history` (bound to `M-r`) and `cape-history` (bound to `C-c p h`). I find tese quite convenient to quickly select from previous commands, instead of cycling through the history. But this is a matter of preference of course.

**Reply by egr95** [reply]
Great, this is all new to me, thanks @minad for the tips!
